# Data consistency check 

### Importing libraries and data sets

### Data consistency check

### Creating Data Frame

### Task 4.5. 

### # Export new dataframe as orders_products_combined


# 1. Importing libraries and data sets

In [4]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [5]:
#creating pathway for dataframes
path = os.path.join(r'/Users/katja/Documents/Instacart Basket Analysis/02 Data')

In [6]:
#import products.csv
df_prods = pd.read_csv(os.path.join(path, 'Original Data','products.csv'), index_col = False)

In [8]:
#import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'),index_col = 0)

# 2. Data consistency check

In [9]:
#removed eval_set column from orders_wrangled.csv
df_ords = df_ords.drop (columns =['eval_set'])

# creating a dataframe

In [10]:
#Create a dataframe
df_test= pd.DataFrame()

In [11]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

#check
df_test.head()

mix
0     a
1     b
2     1
3  True

In [47]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [48]:
#change numeric to string for df_test
df_test['mix'] = df_test['mix'].astype('str')

In [49]:
#checking for missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [50]:
#new df for product Nan values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

#check
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [51]:
df_prods.shape

(49693, 5)

In [52]:
#new products df with no missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

#shape check
df_prods_clean.shape

(49677, 5)

In [53]:
#new df (df_prods_clean) for finding duplicates
df_dups=df_prods_clean[df_prods_clean.duplicated()]

#check
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
df_prods_clean.shape

(49677, 5)

In [55]:
#new df with no duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [56]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [57]:
#Step 2 investigating orders_df - no concerns found
df_ords.describe()

order_id       user_id  clients_order_number     order_dow  \
count  3.421083e+06  3.421083e+06          3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05          1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04          1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00          1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04          5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05          1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05          2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05          1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [58]:
#Step 3 Checking for mixed type data for df_ords - no issues, mixed-type column found
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [59]:
#Step 5 Missing Values check - 
df_ords.isnull().sum()

order_id                       0
user_id                        0
clients_order_number           0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Step 6
#### A substantial amount of data, totaling 206,209 observations, contains missing values. Deleting or imputing these values is deemed risky, as it may introduce inaccuracies into the analysis. Specifically, the dataframe df_ords_missing_values exhibits null values across all observations where order_number equals 1 for each user. This aligns with the logical expectation that the first order for every user may lack values in the days_since_prior_order column. Given that there are 206,209 users, it follows that each user will have a single NaN value corresponding to their initial order. Considering these considerations, the decision has been made to preserve the integrity of the data by leaving the missing values untouched. Maybe in the next step all the orders = 1 can be flagged as new_customer

In [63]:
#df_ords dataframe has 3421083 observations including null values in days_since_prior_order

df_ords.shape

(3421083, 6)

In [64]:
#extract all the missing value observations to a new dataframe (subset)

df_ords_missing_values=df_ords[df_ords['days_since_prior_order'].isnull() == True]

df_ords_missing_values

order_id  user_id  clients_order_number  order_dow  \
0         2539329        1                     1          2   
11        2168274        2                     1          2   
26        1374495        3                     1          1   
39        3343014        4                     1          6   
45        2717275        5                     1          3   
...           ...      ...                   ...        ...   
3420930    969311   206205                     1          4   
3420934   3189322   206206                     1          3   
3421002   2166133   206207                     1          6   
3421019   2227043   206208                     1          1   
3421069   3154581   206209                     1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [62]:
#new df (df_prods_clean) for finding duplicates 
df_dups_ords = df_ords[df_ords.duplicated()]

#check - no duplicates found 
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, clients_order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

### Step 8
#### An empty df_ords_duplicates dataframe proves there are no duplicates.  No action needed to remove duplicates.

# Export Data Sets

In [65]:
#exporting cleaned products - df_prods_clean_no_dups to prepared data folder as a new file products_checked.csv
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 'products_checked.csv'), index=False)

In [66]:
#exporting cleaned ords - df_ords to prepared data folder as a new file products_checked.csv
df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_checked.csv'), index=False)